### ISSUE TITLE
CURLMOPT_SOCKETFUNCTION callback gets called with socket = 0

### ISSUE THREAD

**Issue by bsnote** [curl/curl#1111]
### I did this
I followed the multi-uv example (https://github.com/curl/curl/blob/master/docs/examples/multi-uv.c). Sometimes the `curl_multi_socket_action(curl_handle, CURL_SOCKET_TIMEOUT, 0, &running_handles)` call in `on_timeout()` function results in `handle_socket()` being called with `s = 0`. It starts watching sockfd = 0 and crashes.

### I expected the following

My expectation is that `handle_socket()` (CURLMOPT_SOCKETFUNCTION callback) is always called with a valid socket:

```
int handle_socket(CURL *easy, curl_socket_t s, int action, void *userp, void *socketp) {
  assert(s != 0);
  ...
}
```

Is my assumption correct?
### curl/libcurl version

7.51.0

### operating system

Linux 40:E7:93:0C:53:37 3.1.10 #*15* PREEMPT Tue Mar 18 20:50:32 CST 2014 mips GNU/Linux

**Comment by bsnote** [comment]
As far as I can see, this branch is executed: https://github.com/curl/curl/blob/curl-7_51_0/lib/multi.c#L2615 and then it enters this loop: https://github.com/curl/curl/blob/curl-7_51_0/lib/multi.c#L2623 where it finds an easy handle having `easy_conn->sockfd = 0` and calls `singlesocket()` here: https://github.com/curl/curl/blob/curl-7_51_0/lib/multi.c#L2635

Should "expire-times left to process" loop (https://github.com/curl/curl/blob/curl-7_51_0/lib/multi.c#L2623-L2647) be executed in case `curl_multi_socket_action()` is called with `CURL_SOCKET_TIMEOUT`? Or is there a `return` statement missing in this branch: https://github.com/curl/curl/blob/curl-7_51_0/lib/multi.c#L2610-L2616 ?

**Comment by bsnote** [comment]
cc @bagder

**Comment by bagder** [comment]
> It starts watching sockfd = 0 and crashes.

Why does it crash because of that? Zero is a valid file descriptor. Most commonly used for stdin.

**Comment by bsnote** [comment]
Yes, it is a valid file descriptor. It wasn't added to multi handle though. The application is working with HTTP transfers only. So it's not quite clear where it came from.

**Comment by bagder** [comment]
But you said it crashed. Didn't it?

**Comment by bsnote** [comment]
Yes, it crashed in libuv's `uv_poll_start()` call chain. It looks like a secondary issue though. I assume that `s = 0` is invalid argument for an http transfer, isn't it? It looks like an uninitialized value which is exposed in error.

**Comment by bsnote** [comment]
Here is the easy handle:

```
#3  0x0060ddd4 in datasync::CurlUv::HandleSocket (easy=0xfaeb60, s=0, what=2, userp=0x0, socketp=0x0)
#4  0x00624814 in singlesocket (multi=0xe867f0, data=0xfaeb60) at /home/ec2-user/app/lib/thirdparty/libcurl/lib/multi.c:2347
2347          multi->socket_cb(data,
Current language:  auto; currently c
(gdb) p data
$1 = (struct Curl_easy *) 0xfaeb60
(gdb) p *data
$2 = {next = 0x0, prev = 0x0, easy_conn = 0xd9c390, mstate = CURLM_STATE_WAITCONNECT, result = CURLE_OK, msg = {extmsg = {msg = CURLMSG_DONE, easy_handle = 0xfaeb60, 
      data = {whatever = 0x0, result = CURLE_OK}}}, sockets = {40, 0, 0, 0, 0}, numsocks = 0, dns = {hostcache = 0xe8681c, hostcachetype = HCACHE_MULTI}, 
  multi = 0xe867f0, multi_easy = 0x0, share = 0x0, req = {size = 0, bytecountp = 0x0, maxdownload = -1, writebytecountp = 0x0, bytecount = 0, writebytecount = 0, 
    headerbytecount = 0, deductheadercount = 0, start = {tv_sec = 388, tv_usec = 389423}, now = {tv_sec = 388, tv_usec = 389423}, header = true, 
    badheader = HEADER_NORMAL, headerline = 0, hbufp = 0xe8e518 "\r\n", hbuflen = 0, str = 0x0, str_start = 0x0, end_ptr = 0x0, p = 0x0, content_range = false, 
    offset = 0, httpcode = 0, start100 = {tv_sec = 0, tv_usec = 0}, exp100 = EXP100_SEND_DATA, upgr101 = UPGR101_INIT, auto_decoding = 0, timeofdoc = 0, bodywrites = 0, 
    buf = 0xfaf0f4 ..., 
    uploadbuf = 0xfb70f5 ..., maxfd = 0, keepon = 0, upload_done = false, ignorebody = false, ignorecl = false, location = 0x0, newurl = 0x0, upload_present = 0, upload_fromhere = 0x0, 
    chunk = false, upload_chunky = false, getheader = false, forbidchunk = false, protop = 0xe6fc08}, set = {err = 0x2c67c588, debugdata = 0x0, errorbuffer = 0xe8e368 "", 
    proxyport = 1080, out = 0xe8e310, in_set = 0x2c67c448, writeheader = 0xe8e310, rtp_out = 0x0, use_port = 0, httpauth = 1, proxyauth = 1, followlocation = 0, 
    maxredirs = -1, keep_post = 0, free_referer = false, postfields = 0x0, seek_func = 0, postfieldsize = 0, localport = 0, localportrange = 0, fwrite_func = 0x611ce0
     ..., 
    fwrite_header = 0x611e48 ..., fwrite_rtp = 0, 
    fread_func_set = 0x40a7c0 <fread@plt>, is_fread_set = 0, is_fwrite_set = 1, fprogress = 0, fxferinfo = 0, fdebug = 0, ioctl_func = 0, fsockopt = 0, 
    sockopt_client = 0x0, fopensocket = 0, opensocket_client = 0x0, fclosesocket = 0, closesocket_client = 0x0, seek_client = 0x0, convfromnetwork = 0, convtonetwork = 0, 
    convfromutf8 = 0, progress_client = 0x0, ioctl_client = 0x0, timeout = 120000, connecttimeout = 0, accepttimeout = 0, server_response_timeout = 0, tftp_blksize = 0, 
    tftp_no_options = false, filesize = -1, low_speed_limit = 0, low_speed_time = 0, max_send_speed = 0, max_recv_speed = 0, set_resume_from = 0, headers = 0xe8e4e0, 
    proxyheaders = 0x0, httppost = 0x0, sep_headers = true, cookiesession = false, crlf = false, quote = 0x0, postquote = 0x0, prequote = 0x0, source_quote = 0x0, 
    source_prequote = 0x0, source_postquote = 0x0, telnet_options = 0x0, resolve = 0x0, connect_to = 0x0, timecondition = CURL_TIMECOND_NONE, timevalue = 0, 
    httpreq = HTTPREQ_POST, httpversion = 0, ssl = {version = 0, certverifyresult = 0, verifypeer = true, verifyhost = true, verifystatus = false, CApath = 0x0, 
      CAfile = 0xe8e620 ..., CRLfile = 0x0, issuercert = 0x0, clientcert = 0x0, random_file = 0x0, egdsocket = 0x0, 
      cipher_list = 0x0, max_ssl_sessions = 5, fsslctx = 0, fsslctxp = 0x0, sessionid = true, certinfo = false, falsestart = false, username = 0x0, password = 0x0, 
      authtype = CURL_TLSAUTH_NONE}, proxytype = CURLPROXY_HTTP, dns_cache_timeout = 300, buffer_size = 0, private_data = 0xe8e4c0, http200aliases = 0x0, ipver = 0, 
    max_filesize = 0, ftp_filemethod = FTPFILE_MULTICWD, ftp_create_missing_dirs = 0, ssh_keyfunc = 0, ssh_keyfunc_userp = 0x0, printhost = false, get_filetime = false, 
    tunnel_thru_httpproxy = false, prefer_ascii = false, ftp_append = false, ftp_list_only = false, ftp_use_port = false, hide_progress = true, 
    http_fail_on_error = false, http_keep_sending_on_error = false, http_follow_location = false, http_transfer_encoding = false, 
    http_disable_hostname_check_before_authentication = false, include_header = false, http_set_referer = false, http_auto_referer = false, opt_no_body = false, 
    upload = false, use_netrc = CURL_NETRC_IGNORED, verbose = false, krb = false, reuse_forbid = false, reuse_fresh = false, ftp_use_epsv = true, ftp_use_eprt = true, 
    ftp_use_pret = false, use_ssl = CURLUSESSL_NONE, ftpsslauth = CURLFTPAUTH_DEFAULT, ftp_ccc = CURLFTPSSL_CCC_NONE, no_signal = true, global_dns_cache = false, 
    tcp_nodelay = true, ignorecl = false, ftp_skip_ip = false, connect_only = false, ssl_enable_beast = false, ssl_no_revoke = false, ssh_auth_types = -1, 
    http_te_skip = false, http_ce_skip = false, new_file_perms = 420, new_directory_perms = 493, proxy_transfer_mode = false, str = {0x0 <repeats 19 times>, 
      0xe8e6b8 ..., 0x0, 
      0xe8e620 ..., 0x0 <repeats 24 times>}, scope_id = 0, allowed_protocols = -1, redir_protocols = -201327633, mail_rcpt = 0x0, 
    sasl_ir = false, rtspreq = RTSPREQ_OPTIONS, rtspversion = 0, wildcardmatch = false, chunk_bgn = 0, chunk_end = 0, fnmatch = 0, fnmatch_data = 0x0, 
    gssapi_delegation = 0, tcp_keepalive = false, tcp_keepidle = 60, tcp_keepintvl = 60, tcp_fastopen = false, maxconnects = 5, ssl_enable_npn = true, 
    ssl_enable_alpn = true, path_as_is = false, pipewait = false, expect_100_timeout = 1000, stream_depends_on = 0x0, stream_depends_e = false, stream_weight = 0}, 
  change = {url = 0xe8e6b8 ..., url_alloc = false, 
    referer = 0x0, referer_alloc = false, cookielist = 0x0, resolve = 0x0}, cookies = 0x0, progress = {lastshow = 388, size_dl = 0, size_ul = 0, downloaded = 0, 
    uploaded = 0, current_speed = 0, callback = false, width = 0, flags = 16, timespent = 0.00051999999999999995, dlspeed = 0, ulspeed = 0, t_nslookup = 0.000223, 
    t_connect = 0, t_appconnect = 0, t_pretransfer = 0, t_starttransfer = 0, t_redirect = 0, start = {tv_sec = 388, tv_usec = 389229}, t_startsingle = {tv_sec = 388, 
      tv_usec = 389234}, t_startop = {tv_sec = 388, tv_usec = 389233}, t_acceptdata = {tv_sec = 0, tv_usec = 0}, ul_limit_start = {tv_sec = 0, tv_usec = 0}, 
    ul_limit_size = 0, dl_limit_start = {tv_sec = 0, tv_usec = 0}, dl_limit_size = 0, speeder = {0, 0, 541, 541, 0, 0}, speeder_time = {{tv_sec = 388, tv_usec = 389721}, {
        tv_sec = 256, tv_usec = 22272}, {tv_sec = 257, tv_usec = 16312}, {tv_sec = 257, tv_usec = 16351}, {tv_sec = 0, tv_usec = 0}, {tv_sec = 0, tv_usec = 0}}, 
    speeder_c = 1}, state = {conn_cache = 0xe86858, multi_owned_by_easy = false, keeps_speed = {tv_sec = 0, tv_usec = 0}, lastconnect = 0xdda080, 
    headerbuff = 0xe8e518 "\r\n", headersize = 256, 
    buffer = ..., 
    uploadbuffer = ..., '\0' <repeats 32620 times>, current_speed = -1, this_is_a_follow = false, first_host = 0xf09310 ..., first_remote_port = 443, 
    session = 0xe87610, sessionage = 3, tempwrite = 0x0, tempwritesize = 0, tempwritetype = 0, scratch = 0x0, errorbuf = false, os_errno = 0, prev_signal = 0, 
    allow_port = true, digest = {nonce = 0x0, cnonce = 0x0, realm = 0x0, algo = 0, stale = false, opaque = 0x0, qop = 0x0, algorithm = 0x0, nc = 0}, proxydigest = {
      nonce = 0x0, cnonce = 0x0, realm = 0x0, algo = 0, stale = false, opaque = 0x0, qop = 0x0, algorithm = 0x0, nc = 0}, authhost = {want = 1, picked = 0, avail = 0, 
      done = true, multi = false, iestyle = false}, authproxy = {want = 1, picked = 0, avail = 0, done = true, multi = false, iestyle = false}, authproblem = false, 
    resolver = 0xfbf308, engine = 0x0, expiretime = {tv_sec = 388, tv_usec = 389711}, timenode = {smaller = 0x0, larger = 0x0, same = 0x0, key = {tv_sec = 388, 
        tv_usec = 389711}, payload = 0xfaeb60}, timeoutlist = 0xd9a690, most_recent_ftp_entrypath = 0x0, ftp_trying_alternative = false, httpversion = 0, 
    expect100header = false, pipe_broke = false, prev_block_had_trailing_cr = false, crlf_conversions = 0, 
    pathbuffer = 0xd7f940 ..., 
    path = 0xd7f940 ..., slash_removed = false, use_range = false, rangestringalloc = false, 
    range = 0x0, resume_from = 0, rtsp_next_client_CSeq = 0, rtsp_next_server_CSeq = 0, rtsp_CSeq_recv = 0, infilesize = 0, drain = 0, done = false, 
    fread_func = 0x40a7c0 <fread@plt>, in = 0x2c67c448, stream_depends_on = 0x0, stream_depends_e = false, stream_weight = 0}, wildcard = {state = CURLWC_INIT, 
    path = 0x0, pattern = 0x0, filelist = 0x0, tmp = 0x0, tmp_dtor = 0, customptr = 0x0}, info = {httpcode = 0, httpproxycode = 0, httpversion = 0, filetime = -1, 
    timecond = false, header_size = 0, request_size = 0, proxyauthavail = 0, httpauthavail = 0, numconnects = 1, contenttype = 0x0, wouldredirect = 0x0, 
    conn_primary_ip = "\0002.41.96.3", '\0' <repeats 35 times>, conn_primary_port = 0, conn_local_ip = "\0000.10.10.10", '\0' <repeats 34 times>, conn_local_port = 0, 
    certs = {num_of_certs = 0, certinfo = 0x0}}, tsi = {backend = CURLSSLBACKEND_NONE, internals = 0x0}, magic = 3235830701}
```

**Comment by bagder** [comment]
THe `CURLM_STATE_WAITCONNECT` state is interesting info.

You see the callback is called on line 2347.

If you check before that within the same function, on line 2309, that's the function that populates the sock[] array with sockets this handle is using in this state. It calls `multi_getsock`, which in turn calls the correct subfunction based on the state. `waitconnect_getsock` should've been called in the waitconnect state. It gets the sockets from `conn->tempsock[i]` ... and thus clearly one of those had to be a zero.

The question is then (assuming that my logic is right): how come?

**Comment by bsnote** [comment]
My understanding is that there is nothing unusual in the state of that easy handle. I guess there is no socket connection in that state yet, so there is no valid socket to be assigned. So why is the socket callback called if there is no socket yet?

**Comment by bagder** [comment]
WAITCONNECT implies it waits for the (TCP) connection to complete. In that phase it can have one or two sockets to monitor (perhaps two because of happy eyeballs).

**Comment by bsnote** [comment]
Why does it call the socket callback in that state (`data->mstate = CURLM_STATE_WAITCONNECT`, `data->easy_conn->sockfd = 0`)? Is this the expected behavior?

**Comment by bagder** [comment]
I thought we agreed that file descriptor zero probably is an error there. It calls the socket callback because it detected that the socket set changed so it has to inform the application.

The question is why the `waitconnect_getsock()` function gets socket 0 back. The fact that `data->easy_conn->sockfd` also contains zero is probably a symptom of the same problem. That's at least my guess.

Is it hard for you to trigger this problem? Do you also get the problem on a plain x86 linux? Is it related to a specific URL/host or do you see it with different ones?

**Comment by bsnote** [comment]
Finally I've tracked down the problem. The fd 0, which was initially assigned to a pipe, was unexpectedly closed due to a bug in our application. After the close it is immediately assigned to a socket for subsequent HTTP transfer as it is the smallest available fd. Then it is unexpectedly closed again and this leads to appearing of the second watcher for the same fd, which results in `loop->watchers[w->fd] == w` assertion failure in libuv.

Thanks Daniel for your prompt help and patience! Closing the issue.
